### DISCUSSION TITLE
Off-chain orderbook/swaps

### DISCUSSION THREAD

**Discussion by technicallyty** [regen-network/regen-ledger#585] [Ideas]
It would be neat to have the ability to setup an off-chain way to conduct trades (finalized on chain) similar to the [0x Project](https://0x.org/) and [Sudo Swap](https://sudoswap.xyz/#/). Sudo Swap is built using the 0x project and allows people to create trades for any asset (a basket of ERC20's + a basket of ERC721 for someone else's combination of basket assets). Once the trade it set up, it generates a link for you to send to the counter party who can then accept the trade, and it will be relayed by 0x relayers. The links can be made to be filled by only a specific address or any arbitrary address. 

Something like this on regen ledger (or cosmos as a whole) would be cool, allowing sellers to conduct more specific trades (i.e. 3 carbon credits for your NFT + a bundle of tokens), and have something as simple as a generated link to send to the counter party.

**Comment by aaronc** [comment]
Thanks for sharing this idea @technicallyty. How would the on-chain functionality to support this look?

**Reply by technicallyty** [reply]
I'd imagine we'd just need a mechanism for a tx bundle, such as a `MsgSend` and `MsgSendCredits` bundle with both signatures over the bundle in order to execute. I'm unsure if the sdk supports such bundles natively but perhaps we can make special messages for these batches

**Reply by aaronc** [reply]
We can batch msg's natively in a tx. You'd just need both parties to sign the tx. It should be totally possible now if that's all that's needed

**Reply by technicallyty** [reply]
> You'd just need both parties to sign the tx

So would this allow us to concatenate two messages and put a signature over the combined bytes? the main issue im thinking of right now is a user signing msg1, and then the other user can just submit it and then the trade is busted

**Reply by aaronc** [reply]
You'd need two messages in `TxBody`. Both signers would need to sign over both of them. You can't decompose the signature to just execute one of the messages. The transaction either succeeds or fails atomically. Does that make sense?
